src/S3Persistor.js

```diff
@@ -0,0 +1,366 @@
+
+async function sendFile(bucketName, key, fsPath) {
+  return sendStream(bucketName, key, fs.createReadStream(fsPath))
+}
+
+async function sendStream(bucketName, key, readStream, sourceMd5) {
+  try {
+    // egress from us to S3
+    const observeOptions = { metric: 's3.egress' }
+    let b64Hash
+
+    if (sourceMd5) {
+      b64Hash = PersistorHelper.hexToBase64(sourceMd5)
+    } else {
+      // if there is no supplied md5 hash, we calculate the hash as the data passes through
+      observeOptions.hash = 'md5'
+    }
+
+    const observer = new PersistorHelper.ObserverStream(observeOptions)
+    pipeline(readStream, observer)
+
+    // if we have an md5 hash, pass this to S3 to verify the upload
+    const uploadOptions = {
+      Bucket: bucketName,
+      Key: key,
+      Body: observer
+    }
+    if (b64Hash) {
+      uploadOptions.ContentMD5 = b64Hash
+    }
+
+    const response = await _getClientForBucket(bucketName)
+      .upload(uploadOptions, { partSize: Settings.s3.partSize })
+      .promise()
+    let destMd5 = _md5FromResponse(response)
+    if (!destMd5) {
+      // the eTag isn't in md5 format so we need to calculate it ourselves
+      const verifyStream = await getObjectStream(
+        response.Bucket,
+        response.Key,
+        {}
+      )
+      destMd5 = await PersistorHelper.calculateStreamMd5(verifyStream)
+    }
+
+    // if we didn't have an md5 hash, we should compare our computed one with S3's
+    // as we couldn't tell S3 about it beforehand
+    if (!sourceMd5) {
+      sourceMd5 = observer.getHash()
+      // throws on mismatch
+      await PersistorHelper.verifyMd5(
+        S3Persistor,
+        bucketName,
+        key,
+        sourceMd5,
+        destMd5
+      )
+    }
+  } catch (err) {
+    throw PersistorHelper.wrapError(
+      err,
+      'upload to S3 failed',
+      { bucketName, key },
+      WriteError
+    )
+  }
+}
+
+async function getObjectStream(bucketName, key, opts) {
+  opts = opts || {}
+
+  const params = {
+    Bucket: bucketName,
+    Key: key
+  }
+  if (opts.start != null && opts.end != null) {
+    params.Range = `bytes=${opts.start}-${opts.end}`
+  }
+
+  const stream = _getClientForBucket(bucketName)
+    .getObject(params)
+    .createReadStream()
+
+  // ingress from S3 to us
+  const observer = new PersistorHelper.ObserverStream({ metric: 's3.ingress' })
+
+  try {
+    // wait for the pipeline to be ready, to catch non-200s
+    await PersistorHelper.getReadyPipeline(stream, observer)
+    return observer
+  } catch (err) {
+    throw PersistorHelper.wrapError(
+      err,
+      'error reading file from S3',
+      { bucketName, key, opts },
+      ReadError
+    )
+  }
+}
+
+async function getRedirectUrl() {
+  // not implemented
+  return null
+}
+
+async function deleteDirectory(bucketName, key) {
+  let response
+
+  try {
+    response = await _getClientForBucket(bucketName)
+      .listObjects({ Bucket: bucketName, Prefix: key })
+      .promise()
```

Comment:
`listObjects()` does not handle pagination the way its GCS counterpart does. Each call returns 1000 objects, so we may not have hit that limitation, but it's an easy trap to fall into.